Sources/CompressNIO/zlib/ByteBuffer+zlib+compress.swift

```diff
@@ -0,0 +1,218 @@
+
+import NIOCore
+
+// compress extensions to ByteBuffer
+extension ByteBuffer {
+    /// Compress the readable contents of this byte buffer into another using the compression algorithm specified
+    /// - Parameters:
+    ///   - buffer: Byte buffer to write compressed data to
+    ///   - algorithm: Algorithm to use when compressing
+    /// - Throws:
+    ///     - `NIOCompression.Error.bufferOverflow` if output byte buffer doesnt have enough space to
+    ///         write the compressed data into
+    public mutating func compress(to buffer: inout ByteBuffer, with algorithm: ZlibAlgorithm, configuration: ZlibConfiguration = .init()) throws {
+        var compressor = ZlibCompressor(algorithm: algorithm, configuration: configuration)
+        try compressor.deflate(from: &self, to: &buffer)
+    }
+
+    /// Allocate a new byte buffer and compress this byte buffer into it using the compression algorithm specified
+    /// - Parameters:
+    ///   - algorithm: Algorithm to use when compressing
+    ///   - allocator: Byte buffer allocator used to create new byte buffer
+    /// - Returns: the new byte buffer with the compressed data
+    public mutating func compress(
+        with algorithm: ZlibAlgorithm,
+        configuration: ZlibConfiguration = .init(),
+        allocator: ByteBufferAllocator = ByteBufferAllocator()
+    ) throws -> ByteBuffer {
+        var compressor = ZlibCompressor(algorithm: algorithm, configuration: configuration)
+        var buffer = allocator.buffer(capacity: compressor.maxSize(from: self))
+        try compressor.deflate(from: &self, to: &buffer)
+        return buffer
+    }
+
+    /// A version of compressStream which you provide a fixed sized window buffer to and a process closure.
+    ///
+    /// When the window buffer is full the process closure is called. If there is any unprocessed data left
+    /// at the end of the compress the process closure is called with this.
+    ///
+    /// Before calling this you need to provide a working window `ByteBuffer` to the compressor by setting
+    /// `NIOCompressor.window`.
+    ///
+    /// - Parameters:
+    ///   - compressor: Algorithm to use when compressing
+    ///   - flush: how compressor should flush output data.
+    ///   - process: Closure to be called when window buffer fills up or compress has finished
+    /// - Returns: `ByteBuffer` containing compressed data
+    public mutating func compressStream(
```

Comment:
Can we make this work with an AsyncSequence as in- and/or output?

Reply:
I can review some of the APIs but as the Compressor is non-copyable we have to be clearer about who owns it. So using it in an AsyncSequence is problematic. Eg I have a compressor I want to re-use, if I create an AsyncSequence that takes ownership of it there is no easy to pass it back once iteration has completed.

Reply:
Hmm, unfortunate.